services/web/src/modals/EditProduct.js

```diff
@@ -9,17 +9,17 @@ import UploadsField from 'components/form-fields/Uploads';
 import CurrencyField from 'components/form-fields/Currency';
 // --- Generator: end
 
-export default class EditProduct extends React.Component {
+function getDefaultState(props = {}) {
```

Comment:
why not have this as a method on the class?

also what do you think of this... I think might a more standard React way to handle this might be something like

```js

constructor(props) {
  this.state = {
    open: false,
    error: null,
    loading: false,
  };
}

componentDidUpdate(lastProps, lastState) {
  if (lastState.open !== this.state.open) {
    this.setState({
      error: null,
      product: this.props.product || {},
    });
  }
}
```

This should be able to handle all scenarios... when it mounts it will not be open, when it is opened state will update to whatever current initialized state is and it could all exist in one place without the need for another method

Reply:
alternately:

```js
componentDidUpdate(lastProps, lastState) {
  if (this.state.open && !lastState.open) {
    this.setState({
      error: null,
      product: this.props.product || {},
    });
  }
}
```

this would only initialize when it is open as there is no need to update state on close

Reply:
note I'm leaving out loading intentionally as that bit of state doesn't seem to relate to the open/closed state of the modal... arguably error doesn't as well, however when initializing I suppose you don't want previous error messages to appear

Reply:
Having it outside for the class just because it require doesn't have state. 

I would argue that duplicating what the initialState is bad, but i can still move into a function on the class, no biggie.

I normally stay away from componentDidUpdate because its gets very easy complicated. By far the most error prone code happens here. 

I would argue that just calling onClose is much simpler to follow. Than knowing there is a hook that triggers if the state changes, in different part of the class.

Reply:
> I would argue that duplicating what the initialState is bad

I don't really understand this one... what's the difference between `getDefaultState` and any other method on the class? (other than the lifecycle methods)

> By far the most error prone code happens here.

I think if that's true it's because we're not understanding it correctly... `componentDidUpdate` can get complicated but I think it's generally the most correct place to work with the react lifecycle, not against it... if it gets to complicated usually it's an indication that the compontent is doing too much in the first place..  really don't think we should be afraid to use the lifecycle methods

Reply:
to me `componentDidUpdate` is simpler to follow... when the `open` state changes from one thing to another an action should be taken... having an `onClose` method might be acceptable as well but it's not as future proof as when the component grows other things could potentially flip that state, which is why I find the lifecycle methods to be more stable and actually LESS error prone

Reply:
```I don't really understand this one... what's the difference between getDefaultState and any other method on the class? (other than the lifecycle methods)
```
Yeah thats fine, can be inside, np, will update. Was just arguing for having a method to be called to get the initial state, vs having to duplicate what the initial state is in different parts of the code. 

Still not sold on the `componentDidUpdate` but if you feel strong for using it I will go along. I prefer the other solution because it allows me to ignore the componentDidUpdate (well in this case).

Reply:
what if we combined this with `onOpen` then, as this would have the same meaning? ie.

```js

onOpen = () => {
     this.setState({
      open: true,
      error: null,
      product: this.props.product || {},
    });
}

render() {
  return (
    <Modal onOpen={this.onOpen}>
       ....
    </Modal>
  );
}
```

Reply:
Yeah that seems fine
I am unsure if you suggesting we dont have `this.getDefaultState`

e.g.
```
onOpen = () => {
     this.setState(this.getDefaultState());
}
```

I am only thinking that we need a default state to be set (on on first render), to avoid any reference issue, but perhaps i am wrong

Reply:
yeah I think that if we do it that way we shouldn't need getDefaultState as it would all happen in one place which is what I was going for

Reply:
but thats not how it works, to my understanding. 
the render happens before it is open unless we do like below or another construct

```
<modal>
{open && <modal.content ...>}
</modal>
```

Reply:
ah true... well I guess that's one way to do it... not really opposed to either way I guess

Reply:
it's sort of silly that Modal renders it's children even though they don't appear... also that it doesn't handle it's own `open` state... I wonder if we can't create a slightly better `Modal` wrapper and solve this issue at the same time

Reply:
ohh yes i really hate that so it would be nice to solve

Reply:
let me know what you think https://gist.github.com/kaareal/27057e1c0b897911cce744385d7aff2c